<a href="https://colab.research.google.com/github/yasmin26n/Komputasi-Intelegensia/blob/main/TaskWeek7_Yasmin_Nashwa_Alifya.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# TASK WEEK 7

Nama: Yasmin Nashwa Alifya

NPM: 2106725040

Model : https://huggingface.co/ahmedrachid/FinancialBERT-Sentiment-Analysis

In [9]:
!pip install transformers

In [10]:
!pip install emoji==0.6.0

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 51.0/51.0 kB 2.0 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for emoji: filename=emoji-0.6.0-py3-none-any.whl size=49719 sha256=c555299421c9cfa83060a88bd00600ff5bffb57aae26956a546b8a7334009b60
  Stored in directory: /root/.cache/pip/wheels/1b/bd/d9/310c33c45a553798a714e27e3b8395d37128425442b8c78e07
Successfully built emoji


# Model

In [11]:
# Use a pipeline as a high-level helper
from transformers import pipeline

pipe = pipeline("text-classification", model="ahmedrachid/FinancialBERT-Sentiment-Analysis")

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/789 [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/439M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/369 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/226k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/464k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

/usr/local/lib/python3.10/dist-packages/transformers/tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(


In [32]:
pipe("I am extremely happy with the results of this project!")

[{'label': 'neutral', 'score': 0.9533469080924988}]

In [16]:
pipe("Everything feels wrong and not as expected")

[{'label': 'negative', 'score': 0.9870368242263794}]

In [33]:
pipe("If you leave me, I super happy")

[{'label': 'neutral', 'score': 0.9533469080924988}]

# Attention Transformers

In [20]:
from transformers import AutoTokenizer, AutoModelForSequenceClassification
import torch

# Memasukkan token dan model dari Hugging Face
model_name = "ahmedrachid/FinancialBERT-Sentiment-Analysis"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForSequenceClassification.from_pretrained(model_name)

# Prediksi sentimen
def predict_sentiment(text):
    inputs = tokenizer(text, return_tensors="pt")
    outputs = model(**inputs)
    predictions = torch.nn.functional.softmax(outputs.logits, dim=-1)
    labels = ['negative', 'neutral', 'positive']
    predicted_label = labels[torch.argmax(predictions)]
    return predicted_label, predictions

/usr/local/lib/python3.10/dist-packages/transformers/tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(


In [30]:
sentence = "I am extremely happy with the results of this project!"
label, scores = predict_sentiment(sentence)

print(f"Sentence: {sentence}")
print(f"Predicted Sentiment: {label}")

# Mengambil nilai dari scores dan memformatnya
formatted_scores = [f"{score:.8f}" for score in scores[0]]  # Format setiap score

print(f"Confidence Scores: {formatted_scores}")  # Menampilkan scores yang diformat

Sentence: I am extremely happy with the results of this project!
Predicted Sentiment: positive
Confidence Scores: ['0.00006443', '0.00014290', '0.99979264']


In [28]:
sentence = "Everything feels wrong and not as expected"
label, scores = predict_sentiment(sentence)

print(f"Sentence: {sentence}")
print(f"Predicted Sentiment: {label}")


formatted_scores = [f"{score:.5f}" for score in scores[0]]
print(f"Confidence Scores: {formatted_scores}")

Sentence: Everything feels wrong and not as expected
Predicted Sentiment: negative
Confidence Scores: ['0.98704', '0.01265', '0.00032']


In [31]:
sentence = "If you leave me, I super happy"
label, scores = predict_sentiment(sentence)

print(f"Sentence: {sentence}")
print(f"Predicted Sentiment: {label}")


formatted_scores = [f"{score:.5f}" for score in scores[0]]
print(f"Confidence Scores: {formatted_scores}")

Sentence: If you leave me, I super happy
Predicted Sentiment: neutral
Confidence Scores: ['0.00048', '0.95335', '0.04617']


# Kesimpulan

**"I am extremely happy with the results of this project!"**

Hasil prediksi model awal ('Netral', score: 0.9533469080924988) menunjukkan bahwa model tidak cukup mampu menangkap emosi positif yang jelas dalam kalimat ini, meskipun score yang tinggi menunjukkan bahwa ada beberapa indikasi positif. Namun, model baru memberikan hasil yang lebih akurat dengan sentimen ('Positif', score: 0.99979264), menunjukkan keyakinan yang sangat tinggi terhadap ekspresi kebahagiaan. Perubahan ini menandakan bahwa model baru lebih efektif dalam mendeteksi emosi positif yang kuat, menggambarkan bahwa kalimat ini jelas mengekspresikan rasa bahagia.

**"Everything feels wrong and not as expected"**

Hasil prediksi model awal ('Negatif', score: 0.9870368242263794) menunjukkan keyakinan yang sangat tinggi terhadap sentimen negatif dalam kalimat ini. Hal ini mencerminkan dengan jelas bahwa kalimat tersebut mengungkapkan rasa ketidakpuasan atau kekecewaan. Model baru juga memberikan hasil yang konsisten ('Negatif', score: 0.98704). Meskipun ada perbedaan kecil dalam nilai, kedua model secara efektif menangkap emosi negatif yang kuat.

**"If you leave me, I super happy"**

Hasil prediksi model awal ('Netral', score: 0.9533469080924988) menunjukkan bahwa model tidak mampu menangkap ambiguitas emosi dalam kalimat ini, meskipun score yang tinggi menunjukkan adanya ketidakpastian. Meskipun kalimat "I super happy" mengindikasikan perasaan positif, konteks sebelumnya "If you leave me" menciptakan keraguan yang mengaburkan sentimen yang jelas. Model baru memberikan hasil serupa ('Netral', score: 0.95335), yang menunjukkan bahwa meskipun ada indikasi kebahagiaan, model tetap melihat pernyataan ini sebagai netral, mungkin karena ketidakpastian yang ditimbulkan oleh frase pertama yang berpotensi negatif. Ini menunjukkan bahwa kedua model tidak sepenuhnya menangkap kompleksitas emosi yang ada.